src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/UseExceptionThrowHelpersTests.cs

```diff
@@ -0,0 +1,1078 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Xunit;
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Runtime.UseExceptionThrowHelpers,
+    Microsoft.NetCore.Analyzers.Runtime.UseExceptionThrowHelpersFixer>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Runtime.UseExceptionThrowHelpers,
+    Microsoft.NetCore.Analyzers.Runtime.UseExceptionThrowHelpersFixer>;
+
+namespace Microsoft.NetCore.Analyzers.Runtime.UnitTests
+{
+    public class UseExceptionThrowHelpersTests
+    {
+        [Fact]
+        public async Task ArgumentNullExceptionThrowIfNull_DoesntExist_NoDiagnostics()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+
+namespace System
+{
+    public class ArgumentNullException : Exception
+    {
+        public ArgumentNullException(string paramName) { }
+        public ArgumentNullException(string paramName, string message) { }
+    }
+}
+
+class C
+{
+    void M(string arg)
+    {
+        if (arg is null)
+            throw new ArgumentNullException(nameof(arg));
+
+        if (arg == null)
+            throw new ArgumentNullException(nameof(arg));
+
+        if (null == arg)
+            throw new ArgumentNullException(nameof(arg));
+
+        if (arg is null)
+        {
+            throw new ArgumentNullException(nameof(arg));
+        }
+
+        if (arg == null)
+        {
+            throw new ArgumentNullException(nameof(arg));
+        }
+
+        if (null == arg)
+        {
+            throw new ArgumentNullException(nameof(arg));
+        }
+    }
+
+    string this[string name]
+    {
+        get
+        {
+            if (name is null)
+                throw new ArgumentNullException(nameof(name));
+            return name;
+        }
+    }
+}
+");
+        }
+
+        [Fact]
+        public async Task ArgumentNullExceptionThrowIfNull_FixAppliesAsAppropriate()
+        {
+            await new VerifyCS.Test()
+            {
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp9,
+                TestCode =
+@"
+using System;
+
+namespace System
+{
+    public class ArgumentNullException : Exception
+    {
+        public ArgumentNullException() { }
+        public ArgumentNullException(string paramName) { }
+        public ArgumentNullException(string paramName, string message) { }
+        public ArgumentNullException(string paramName, Exception innerException) { }
+        public static void ThrowIfNull(object argument, string name = null) { }           
+    }
+}
+
+class C
+{
+    void M(string arg)
+    {
+        {|CA1858:if (arg is null)
+            throw new ArgumentNullException(nameof(arg));|}
```

Comment:
Consider having a test where the parameter name passed is wrong.

Reply:
Sure.  In such cases, I made the conscious decision to ignore the parameter name being passed to the exception and still "fix" it to use the argument being checked.  It's possible that'll result in the wrong fix being applied, but I've seen enough cases where the argument was accidentally wrong that it felt like the better default.  We can change it to respect exactly what was written if it proves problematic.